merlin/systems/dag/runtimes/nvtabular/executor.py

```diff
@@ -0,0 +1,305 @@
+        if workflow_node.selector:
+            selector_columns = workflow_node.selector.names
+            to_remove = []
+            for upstream_tensors in upstream_outputs or []:
+                for col in selector_columns:
+                    if col in upstream_tensors:
+                        to_remove.append(col)
+            for col in set(to_remove):
+                selector_columns.remove(col)
+
+            if selector_columns:
+                selected_tensors = {c: input_tensors[c] for c in selector_columns}
+                upstream_outputs.append(selected_tensors)
+
+        return upstream_outputs
+
+    def _standardize_formats(self, workflow_node, node_input_data):
+        # Get the supported formats
+        op = workflow_node.op
+        if op and hasattr(op, "inference_initialize"):
+            supported_formats = _maybe_mask_cpu_only(op.supports, self.device)
+        else:
+            supported_formats = Supports.CPU_DICT_ARRAY
+
+        # Convert the first thing into a supported format
+        tensors = _convert_format(node_input_data[0], supported_formats)
+        target_format = _data_format(tensors)
+
+        # Convert the whole list into the same format
+        formatted_tensors = []
+        for upstream_tensors in node_input_data:
+            upstream_tensors = _convert_format(upstream_tensors, target_format)
+            formatted_tensors.append(upstream_tensors)
+
+        return formatted_tensors
+
+
+def _concat_tensors(tensors):
+    format_ = _data_format(tensors[0])
+
+    if format_ & (Supports.GPU_DATAFRAME | Supports.CPU_DATAFRAME):
+        return concat_columns(tensors)
+    else:
+        output = tensors[0]
+        for tensor in tensors[1:]:
+            output.update(tensor)
+        return output
+
+
+def _maybe_mask_cpu_only(supported, device):
+    # if we're running on the CPU only, mask off support for GPU data formats
+    if device == "CPU":
+        supported = functools.reduce(
+            lambda a, b: a | b,
+            (v for v in list(Supports) if v & supported and "CPU" in str(v)),
+        )
+
+    return supported
+
+
+def _get_unique(cols):
+    # Need to preserve order in unique-column list
+    return list({x: x for x in cols}.keys())
+
+
+def _data_format(transformable):
+    data = TensorTable(transformable) if isinstance(transformable, dict) else transformable
+
+    if cudf and isinstance(data, cudf.DataFrame):
+        return Supports.GPU_DATAFRAME
+    elif pandas and isinstance(data, pandas.DataFrame):
+        return Supports.CPU_DATAFRAME
+    elif data.column_type is CupyColumn:
+        return Supports.GPU_DICT_ARRAY
+    elif data.column_type is NumpyColumn:
+        return Supports.CPU_DICT_ARRAY
+    else:
+        if isinstance(data, TensorTable):
+            raise TypeError(f"Unknown type: {data.column_type}")
+        else:
+            raise TypeError(f"Unknown type: {type(data)}")
+
+
+def _convert_format(tensors, target_format):
```

Comment:
A next step along the road of this refactor is to replace `DictArray` with `TensorTable` and get rid of the manual conversions below. I think we'll still need a function like `_convert_format`, but the actual conversions it uses should be the dataframe/`TensorTable` conversions from Merlin Core.